[BREAKING] Disallow legacy constructor

erak opened the pull request:
Closes #4106.

axic commented:
Sorry I forgot to assign the issue to myself, but did mention the last call I started this work. Anyway it is not a big amount of duplicate effort :)

However we should only merge this after #4208

erak commented:
@axic Oh sorry, did not get this in the last call. We should try to keep the project board up to date ;) 
So, should we use with this branch / PR?

I agree, let's merge the other PR first.

axic commented:
@bit-shift I will have a look at the differences between the two PRs. Do you want to review #4208 ?

axic commented:
My local version of these changes are all included int this PR.

erak commented:
@christianparpart I removed the expected shadowing warning of ``override_constructor.sol`` and ``constructor_old.sol``. The old version of ``override_constructor.sol`` was basically testing that you can have a function that is called like the constructor of a contracts base contract. Since this construct is not possible anymore, we also should not warn anymore.
And getting the warning if `` contract A { function A() public {} }``  feels a little bit like clutter.

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -210,36 +210,32 @@ bool SyntaxChecker::visit(PlaceholderStatement const&)
+bool SyntaxChecker::visit(ContractDefinition const& _contract)
+{
+	ASTString const& contractName = _contract.name();
+	for (const FunctionDefinition* const function : _contract.definedFunctions())
```
Should be `FunctionDefinition const* const function`, but I think the last `const` does not really help readability and thus could be dropped.

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -210,36 +210,32 @@ bool SyntaxChecker::visit(PlaceholderStatement const&)
+	for (const FunctionDefinition* const function : _contract.definedFunctions())
+		if (function->name() == contractName)
+			m_errorReporter.syntaxError(function->location(),
+			"Defining constructors as functions with the same name as the contract is not allowed. "
```
Please use the original `v050` message and not the message of the pre-050 warning. Also, the indentation and line breaks are not according to coding style.

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -210,36 +210,32 @@ bool SyntaxChecker::visit(PlaceholderStatement const&)
 			m_errorReporter.warning(_function.location(), "Modifiers of functions without implementation are ignored." );
 	}
-	if (_function.name() == "constructor")
+	if (_function.name() == "constructor" && !_function.isConstructor())
```
Please don't add `!_function.isConstructor()`. Currently, it looks like this is a good thing to add, but it might hide errors that are introduced when we change the logic of `isConstructor`.

chriseth commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -333,30 +330,28 @@ StateMutability Parser::parseStateMutability(Token::Value _token)
 
-	if (result.isConstructor || _forceEmptyName || m_scanner->currentToken() == Token::LParen)
+	if (result.isConstructor)
+		result.name = make_shared<ASTString>("constructor");
```
I think the name should still be empty.

chriseth commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -333,30 +330,28 @@ StateMutability Parser::parseStateMutability(Token::Value _token)
+		result.name = make_shared<ASTString>("constructor");
+	else if (_forceEmptyName || m_scanner->currentToken() == Token::LParen)
 		result.name = make_shared<ASTString>();
+	else if (m_scanner->currentToken() == Token::Constructor && !result.isConstructor)
```
Is this to still allow `function constructor(...) { ..}`? I don't think we have to support that - `constructor` is now a real keyword and thus functions cannot be named `constructor`.

erak replied:
Yes, the idea was to still allow it, but warn about it's usage. But you're right, since this is keyword now, we should not allow it anymore.

chriseth replied:
`!result.isConstructor` can be removed (it is an `else if`).

chriseth commented on test/libsolidity/syntaxTests/constructor/overriding_constructor.sol:
```diff
@@ -1,6 +1,3 @@
-// It is fine to "override" constructor of a base class since it is invisible
-contract A { function A() public { } }
+contract A { constructor() public {} }
 contract B is A { function A() public pure returns (uint8) {} }
```
Yep, that is exactly what we should test here!

chriseth reviewed (changes requested):
Is there a test for
```solidity
contract C {
function constructor() { ... }
}
```
?

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -2656,8 +2656,8 @@ BOOST_AUTO_TEST_CASE(value_for_constructor)
-			function Main() payable {
-				h = (new Helper).value(10)("abc", true);
+			constructor() payable {
+				h = (new Helper).value(10)('abc', true);
```
Any reason for this change?

erak replied:
Oh, this was an unconscious change due to broken code highlighting in my editor. We can revert to ``"abc"``

chriseth commented on test/libsolidity/SolidityParser.cpp:
```diff
@@ -493,6 +493,7 @@ BOOST_AUTO_TEST_CASE(keyword_is_reserved)
 		"after",
 		"case",
 		"catch",
+		"constructor"
```
`constructor` should not be reserved, it should be a keyword "in use".

axic commented:
Depends on #4206.

axic commented:
This could be fixed without the external tests:
```
ASSERTION FAILURE:
- file   : SolidityABIJSON.cpp
- line   : 49
- message: Parsing contract failed
```

Also syntax test expectations need to be updated.

axic commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -333,30 +330,31 @@ StateMutability Parser::parseStateMutability(Token::Value _token)
 		result.name = make_shared<ASTString>();
+	else if (_forceEmptyName || m_scanner->currentToken() == Token::LParen)
+		result.name = make_shared<ASTString>();
+	else if (m_scanner->currentToken() == Token::Constructor && !result.isConstructor)
```
I don't think this can happen since `expectIdentifierToken` shouldn't allow a non-identifier, such as `Token::Constructor`. Is there a test for this which needs this extra condition here?

erak replied:
@axic This check should have been removed already. Thanks!

erak commented:
@axic I updated the broken ABI tests.

erak commented:
@chriseth The file ``function_named_constructor.sol`` contains a test for 

``contract C { function constructor() { ... }}``

chriseth commented:
Please rebase to trigger the tests.

axic commented:
solc-js needs to be updated (thedao code). Also please try to update gnosis first (using their `development` branch), but if it seems to be too much work we can disable it.

Also needs a rebase.

axic commented on Changelog.md:
```diff
@@ -13,6 +13,7 @@ Breaking Changes:
    ``implements``, ``macro``, ``mutable``, ``override``, ``partial``, ``promise``, ``reference``, ``sealed``,
    ``sizeof``, ``supports``, ``typedef`` and ``unchecked``.
  * General: Remove assembly instruction aliases ``sha3`` and ``suicide``
+ * General: Disallow old constructor syntax.
```
Need to extend the "New keywords" line above with `constructor`.

axic commented:
Isn't there a documentation section to be updated/extended after this change?

chriseth commented:
I thought gnosis was not worth keeping because of truffle versioning problems?

axic commented:
Gnosis has two branches, `master` and `development`, the latter uses the new truffle, that should be close to working.

erak commented:
@axic Does that mean that we're not going use your Gnosis fork for now? I've been checking out their ``development`` branch, but it also needs an update.

erak commented:
@axic Regarding documentation: a lot of it was already updated with PR #3635, but I will read through the important sections again.

axic commented:
> Does that mean that we're not going use your Gnosis fork for now? I've been checking out their development branch, but it also needs an update.

You can push their dev branch to our fork's `solidity-050` branch. And we can take it from there :)

bit-shift commented:
I created a PR in ``solc-js`` that reflects the new constructor syntax: ethereum/solc-js#225

bit-shift commented:
- [ ] Switch back ``solc-js`` and Gnosis repositories after ethereum/solc-js#225 and axic/pm-contracts#2 got merged.

erak commented:
@axic We only have an issue with Truffle left, since the version used by ``npm`` (4.1.11) fails while testing the contracts that are been delivered with it. Do you have any information on how their updating schedule looks like? I would probably create a PR that upadates the contracts in https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-core/lib/testing, but don't let it block this one.

TODO:
- [ ] Switch back ``solc-js`` and Gnosis repositories in ``externalTests.sh`` after ethereum/solc-js#225 and axic/pm-contracts#2 got merged.

axic commented:
@bit-shift to avoid a lot of rebasing work while this is still pending, we could merge updating our local test suite to use `constructor` in a separate PR right now?

erak commented:
@axic After 2 unsuccessful attempts to disable the failing Truffle tests, I decided to remove the Gnosis repository from our external tests for now. I will create an issue to remind ourselves to re-add them after the next Truffle release (which will hopefully contain the changes from the PR that I'm going to create within the next days.) I forked and changed the contract that's breaking the test already: https://github.com/bit-shift/truffle/tree/updateConstructor

axic commented:
@bit-shift can you open a PR on truffle? They should be quick enough to merge.

axic commented:
Also see my patch in #3534 in the `externalTests.sh` to change truffle. This could be added that way.

erak commented:
@axic Great! I will apply that workaround.

chriseth commented:
Needs to be rebased.

chriseth commented:
Sorry, but this is just gigantic. Could you try to extract updating the external files into a separate PR first? Then we can focus on the semantics changes.

erak commented:
@chriseth I pulled out all updates to the external compilation tests into a separate PR: #4350, but this one is still huge. I could also extract the tests in ``test/libsolidity`` since it's just the syntax that's been updated in these files and I'd like to keep the actual semantic change and the syntax tests as part of one PR.

erak committed:
Disallows old constructor syntax.

erak commented:
Updated.

chriseth commented:
Tests are failing.

erak commented:
We need to merge https://github.com/ethereum/solidity/pull/4354 first. After merging, this PR needs to be updated such that it removes the old syntax tests that expect a warning instead of an error.

erak committed:
Adjusts syntax tests to new constructor syntax.

erak commented:
Updated this PR again after #4354 got merged. Should be ready for final review now.

erak commented:
I missed the docs here, but will add these changes.

erak commented:
I've split out the required changes to the documentation: https://github.com/ethereum/solidity/pull/4402. After that PR got merged, we can merge this one.

ekpyron commented on libsolidity/analysis/NameAndTypeResolver.cpp:
```diff
@@ -476,7 +476,13 @@ bool DeclarationRegistrationHelper::registerDeclaration(
+				dynamic_cast<FunctionDefinition const*>(&_declaration) &&
+				name == decl->name())
+			)
+				shadowedDeclaration = decl;
```
So only something that is a function can be shadowed by something that's not a contract with the same name?
What about ``contract C { uint msg; }``? Currently this warns: ``This declaration shadows a builtin symbol.``. And I think this comes from below in this function (I may be wrong) - does that still work? Is there a test for it?

erak replied:
We have a test for the case you've mentioned: ``452_shadowing_builtins_with_storage_variables.sol``

erak replied:
After we talked about that: I've improved the formatting to make the condition easier to understand 😄

ekpyron reviewed (dismissed):
Looks fine to me now.

chriseth commented on libsolidity/analysis/NameAndTypeResolver.cpp:
```diff
@@ -476,7 +476,15 @@ bool DeclarationRegistrationHelper::registerDeclaration(
 	Declaration const* shadowedDeclaration = nullptr;
 	if (_warnOnShadow && !name.empty() && _container.enclosingContainer())
 		for (auto const* decl: _container.enclosingContainer()->resolveName(name, true, true))
```
Unrelated to this PR, but why do we have a for loop here?

chriseth commented on libsolidity/analysis/NameAndTypeResolver.cpp:
```diff
@@ -476,7 +476,15 @@ bool DeclarationRegistrationHelper::registerDeclaration(
 	if (_warnOnShadow && !name.empty() && _container.enclosingContainer())
 		for (auto const* decl: _container.enclosingContainer()->resolveName(name, true, true))
-			shadowedDeclaration = decl;
+			// Do not warn about functions shadowing a contract.
```
Can you explain the reason for this?

erak replied:
@chriseth Without this change, the following code would result in a warning about ``function A()`` shadowing ``contract A``:
```
contract A { constructor() public {} }
contract B is A { function A() public pure returns (uint8) {} }
```
Since functions can now have the same name as a previously declared contract, this should not warn. We also have a test for that: ``overriding_constructor.sol``.

chriseth replied:
I would say the warning is correct, since it does shadow `A`. Note the following does not work because the function `A` shadows the contract `A`:
```
contract A { function f() public {} }
contract B is A {
  function A() public pure returns (uint8) {}
  function g() public {
    A.f();
  }
}
```

chriseth commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -210,40 +210,32 @@ bool SyntaxChecker::visit(PlaceholderStatement const&)
+bool SyntaxChecker::visit(ContractDefinition const& _contract)
+{
+	ASTString const& contractName = _contract.name();
+	for (FunctionDefinition const* function : _contract.definedFunctions())
```
No space in front of `:`.

erak committed:
Adds warning if function is shadowing a contract.

erak commented:
Rebased and squashed some commits. @chriseth I've also addressed your comment about a function shadowing a contract: https://github.com/ethereum/solidity/pull/4215#discussion_r202017629.

chriseth commented on test/libsolidity/syntaxTests/constructor/constructor.sol:
```diff
@@ -0,0 +1,9 @@
+contract A { 
```
Please move the positive case into its own test.

erak commented:
@chriseth I reverted merging both, the new and the old constructor syntax, into one file. I also rebased the branch.
